clippy_lints/src/call_missing_target_feature.rs

```diff
@@ -0,0 +1,190 @@
+#![allow(clippy::similar_names)]
+use clippy_utils::diagnostics::span_lint_and_then;
+use rustc_hir as hir;
+use rustc_hir::def::Res;
+use rustc_hir::def_id::DefId;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::declare_lint_pass;
+use rustc_span::{sym, Symbol};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks that the caller enables the target features that the callee requires
+    ///
+    /// ### Why is this bad?
+    /// Not enabling target features can cause UB and limits optimization opportunities.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// #[target_feature(enable = "avx2")]
+    /// unsafe fn f() -> u32 {
+    ///     0
+    /// }
+    ///
+    /// fn g() {
+    ///     unsafe { f() };
+    ///     // g does not enable the target features f requires
+    /// }
+    /// ```
+    #[clippy::version = "1.82.0"]
+    pub CALL_MISSING_TARGET_FEATURE,
+    suspicious,
+    "call requires target features that the surrounding function does not enable"
+}
+
+declare_lint_pass!(CallMissingTargetFeature => [CALL_MISSING_TARGET_FEATURE]);
+
+impl<'tcx> LateLintPass<'tcx> for CallMissingTargetFeature {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &hir::Expr<'tcx>) {
+        if !in_external_macro(cx.tcx.sess, expr.span) {
+            let Some(caller_def_id) = caller_def_id(cx, expr) else {
+                return;
+            };
+            let Some(callee_def_id) = callee_def_id(cx, expr) else {
+                return;
+            };
+
+            let caller_target_features = def_id_target_features(cx, caller_def_id);
+            let callee_target_features = def_id_target_features(cx, callee_def_id);
+
+            let missing: Vec<_> = callee_target_features
+                .iter()
+                .filter(|target_feature| !caller_target_features.contains(target_feature))
+                .map(Symbol::as_str)
+                .collect();
+
+            if missing.is_empty() {
+                return;
+            }
```

Comment:
You should be able to use `TyCtxt::codegen_fn_attrs` instead of parsing the attributes. This will also handle any inherited features as well as implied features.